source-code/core/src/lint/createLintRule.ts

```diff
@@ -0,0 +1,56 @@
+import type { LintRule } from "./rule.js"
+
+/**
+ * A utility function to create a lint rule.
+ *
+ * @example
+ * const myRule = createLintRule({ id: "example.rule" }, (args) => {
+ *   // Implement your rule logic here
+ * });
+ *
+ * // using the rule
+ * myRule("error");
+ *
+ * // you can also pass settings to the rule
+ * const myRule2 = createLintRule<{ strict: boolean }>({ id: "example.rule" }, (args) => {
+ *   // Implement your rule logic here
+ * });
+ *
+ * // settings must be provided when using the rule
+ * myRule2("error", { strict: true });
+ */
+export const createLintRule = <Settings extends object | undefined = undefined>(
+	args: { id: LintRule["id"] },
+	setup: (
+		args: Parameters<LintRule["setup"]>[0] & {
```

Comment:
why isn't `Settings` part of `LintRule["setup"]`?

Reply:
1. The linter can't pass settings to rules.
2. Even if, the implementation would become more complex with no benefit. 

The job of `createLintRule`, which wraps the `setup` function of a rule, is to pass custom arguments like `settings` to a rule.   


```ts

rules: [
  // a "manually" written lint rule
  {
    id: "my.lintRule",
    level: "warn", 
    setup: ({ report }) => {
      // developers can directly access settings as desired + the linting implementation can't
      // pass down user-configured settings to a lint rule.   
      if (strict) {
        report()
      }
    }
  }
  // a lint rule that has been created with createLintRule
  missingMessge("error", { strict: true })
]
```

Reply:
ok, got it